nupic/embodied/multitask/value_functions/continuous_dendrite_mlp_q_function.py

```diff
@@ -0,0 +1,121 @@
+#  See the GNU Affero Public License for more details.
+#
+#  You should have received a copy of the GNU Affero Public License
+#  along with this program.  If not, see http://www.gnu.org/licenses.
+#
+#  http://numenta.org/licenses/
+#
+# ------------------------------------------------------------------------------
+"""This modules creates a continuous Q-function network with a Dendrite MLP."""
+
+import torch
+
+from nupic.embodied.multitask.modules import CustomDendriticMLP
+
+class ContinuousDendriteMLPQFunction(CustomDendriticMLP):
+    """Implements a continuous MLP Q-value network.
+
+    It predicts the Q-value for all actions based on the input state. It uses
+    a PyTorch neural network module to fit the function of Q(s, a).
+    """
+
+    def __init__(
+        self,
+        env_spec,
+        num_tasks,
+        input_data, 
+        context_data,
+        hidden_sizes,
+        layers_modulated,
+        num_segments,
+        kw_percent_on,
+        context_percent_on,
+        weight_sparsity,
+        weight_init,
+        dendrite_weight_sparsity,
+        dendrite_init,
+        dendritic_layer_class,
+        output_nonlinearity,
+        preprocess_module_type,
+        preprocess_output_dim
+    ):
+        """Initialize class with multiple attributes.
+
+        Args:
+            env_spec (EnvSpec): Environment specification.
+            **kwargs: Keyword arguments.
+
+        """
+        self.num_tasks = num_tasks
+
+        self.input_data = input_data
+        self.context_data = context_data
+
+        if input_data == "obs":
+            self.input_dim = env_spec.observation_space.flat_dim - self.num_tasks + env_spec.action_space.flat_dim
+        elif input_data == "obs|context":
+            self.input_dim = env_spec.observation_space.flat_dim + env_spec.action_space.flat_dim
+        
+        if context_data == "context":
+            self.context_dim = self.num_tasks
+        elif context_data == "obs|context":
+            self.context_dim = env_spec.observation_space.flat_dim + env_spec.action_space.flat_dim
+
+
+        super().__init__(
+            input_dim=self.input_dim,
+            context_dim=self.context_dim,
+            output_sizes=1,
+            hidden_sizes=hidden_sizes,
+            layers_modulated=layers_modulated,
+            num_segments=num_segments,
+            kw_percent_on=kw_percent_on,
+            context_percent_on=context_percent_on,
+            weight_sparsity=weight_sparsity,
+            weight_init=weight_init,
+            dendrite_weight_sparsity=dendrite_weight_sparsity,
+            dendrite_init=dendrite_init,
+            dendritic_layer_class=dendritic_layer_class,
+            output_nonlinearity=output_nonlinearity,
+            preprocess_module_type=preprocess_module_type,
+            preprocess_output_dim=preprocess_output_dim,
+            preprocess_kw_percent_on=kw_percent_on
+        )
+
+    def forward(self, observations, actions):
```

Comment:
Same comment as before, ifs in forward are never a good idea. I also see some repetition between the code here and the other network class - is there a common ground that could allow the code to be simplified?

Reply:
I agree on the if's statement portion - I can create some extra forward functions. 

Not sure how I could define common ground without making it too complicated/messy. Information only passes through this class and the policy class, and this is where I can break them apart. Plus, q functions take in additional information (actions) that doesn't exist in the policy setting.